src/game/server/entities/turret.cpp

```diff
@@ -0,0 +1,246 @@
+#include "turret.h"
+#include "plasma.h"
+#include "laser.h"
+
+CTurret::CTurret(CGameWorld *pGameWorld, vec2 Pos, int Owner, vec2 Direction, float StartEnergy, int Type)
+: CEntity(pGameWorld, CGameWorld::ENTTYPE_TURRET)
+{
+	m_Pos = Pos;
+	m_Owner = Owner;
+	m_Energy = StartEnergy;
+	m_Dir = Direction;
+	m_StartTick = Server()->Tick();
+	m_Bounces = 0;
+	m_Radius = 15.0f;
+	m_EvalTick = Server()->Tick();
+	m_OwnerChar = GameServer()->GetPlayerChar(m_Owner);
+	m_LifeSpan = Server()->TickSpeed()*g_Config.m_InfTurretDuration;
+	m_WarmUpCounter = Server()->TickSpeed()*g_Config.m_InfTurretWarmUpDuration;
+	m_ReloadCounter = Server()->TickSpeed()*g_Config.m_InfTurretReloadDuration;
+	m_Type = Type;
+
+	m_IDs.set_size(9);
+	for(int i = 0; i < m_IDs.size(); i++)
+	{
+		m_IDs[i] = Server()->SnapNewID();
+	}
+	
+	GameWorld()->InsertEntity(this);
+
+}
+
+CTurret::~CTurret()
+{
+	for(int i = 0; i < m_IDs.size(); i++)
+	{
+		Server()->SnapFreeID(m_IDs[i]);
+	}
+}
+
+void CTurret::Reset()
+{
+    GameServer()->m_World.DestroyEntity(this);
+}
+
+int CTurret::GetOwner() const
+{
+	return m_Owner;
+}
+
+void CTurret::Tick()
+{
+
+	//marked for destroy
+	if(m_MarkedForDestroy) 
+		return;
+	if(!m_OwnerChar) {
+		Reset();
+		return;
+	}
+	if(m_LifeSpan < 0) 
+		Reset();
+	
+	for(CCharacter *pChr = (CCharacter*) GameWorld()->FindFirst(CGameWorld::ENTTYPE_CHARACTER); pChr; pChr = (CCharacter *)pChr->TypeNext()) 
+	{
+		if(!pChr->IsInfected()) continue;
+		if(pChr->GetClass() == PLAYERCLASS_UNDEAD && pChr->IsFrozen()) continue;
+		if(pChr->GetClass() == PLAYERCLASS_VOODOO && pChr->m_VoodooAboutToDie) continue;
+		
+		float Len = distance(pChr->m_Pos, m_Pos);
+		
+		// selfdestruction
+		if(Len < pChr->m_ProximityRadius + 4.0f )
+		{
+			pChr->TakeDamage(vec2(0.f, 0.f), g_Config.m_InfTurretSelfDestructDmg, m_Owner, WEAPON_RIFLE, TAKEDAMAGEMODE_NOINFECTION);
+			GameServer()->CreateSound(m_Pos, SOUND_RIFLE_FIRE);
+			char aBuf[64];
+			str_format(aBuf, sizeof(aBuf), "You destroyed %s's turret!", Server()->ClientName(m_Owner));
+			GameServer()->SendChatTarget(pChr->GetPlayer()->GetCID(), aBuf);
+			GameServer()->SendChatTarget(m_Owner, "A zombie has destroyed your turret!");
+			Reset();
+		}
+	}
+
+	
+	
+	
+	//reduce lifespan
+	m_LifeSpan--;
+	
+	
+	//reloading in progress
+	if(m_ReloadCounter > 0)
+	{
+		m_ReloadCounter--;
+		
+		if(m_Radius > 15.0f) //shrink radius
+		{
+			m_Radius -= m_RadiusGrowthRate;
+			if(m_Radius < 15.0f)
+				m_Radius = 15.0f;
+			
+		}
+		return; //some reload tick-cycles necessary
+		
+	} 
+
+	//Reloading finished, warm up in progress
+	if ( m_WarmUpCounter > 0 ) 
+	{
+			m_WarmUpCounter--;
+			
+			if(m_Radius < 45.0f)
+			{
+				m_Radius += m_RadiusGrowthRate;
+				if(m_Radius > 45.0f)
+					m_Radius = 45.0f;
+			}
+			
+			return; //some warmup tick-cycles necessary
+	}
+	
+	//warmup finished, ready to find target
+	
+
+	//CCharacter *pChr = GameServer()->m_World.ClosestCharacter(m_Pos, 20.0f, 0);

```

Comment:
is this code necessary? If no, please remove at all

Reply:
true